test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestStrictFieldBarriers.java

```diff
@@ -0,0 +1,227 @@
+        final int x;
+
+        A1() {
+            x = 1;
+            super();
+            sharedA1 = this;
+        }
+    }
+
+    static class B1 {
+        @Strict
+        final int x;
+
+        B1() {
+            x = 1;
+            super();
+            sharedB1 = this;
+        }
+
+        B1(boolean unused) {
+            x = 1;
+            super();
+            sharedC1_2 = this;
+        }
+    }
+
+    static class C1 extends B1 {
+        @Strict
+        final int y;
+
+        C1() {
+            y = 1;
+            super(true);
+            sharedC1_1 = this;
+        }
+    }
+
+    static abstract value class E1 {
+        final int x;
+
+        E1() {
+            x = 1;
+            super();
+            sharedD1_2 = this;
+        }
+    }
+
+    static class D1 extends E1 {
+        final int y;
+        final int z;
+
+        D1() {
+            y = 2;
+            super();
+            z = 3;
+            sharedD1_1 = this;
+        }
+    }
+
+    // Non final versions
+
+    static class A2 {
+        @Strict
+        int x;
+
+        A2() {
+            x = 1;
+            super();
+            sharedA2 = this;
+        }
+    }
+
+    static class B2 {
+        @Strict
+        int x;
+
+        B2() {
+            x = 1;
+            super();
+            sharedB2 = this;
+        }
+
+        B2(boolean unused) {
+            x = 1;
+            super();
+            sharedC2_2 = this;
+        }
+    }
+
+    static class C2 extends B2 {
+        @Strict
+        int y;
+
+        C2() {
+            y = 1;
+            super(true);
+            sharedC2_1 = this;
+        }
+    }
+
+    static abstract value class E2 {
+        final int x;
+
+        E2() {
+            x = 1;
+            super();
+            sharedD2_2 = this;
+        }
+    }
+
+    static class D2 extends E2 {
+        final int y;
+        final int z;
+
+        D2() {
+            y = 2;
+            super();
+            z = 3;
+            sharedD2_1 = this;
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        // Spawn two threads, a reader and a writer and check that the
+        // reader thread never observes an unitialized strict field.
+        Thread reader = new Thread(() -> {
+            for (int i = 0; i < 100_000; ++i) {
+                // We don't check individual fields here because the checks need to be
+                // as fast a possible to increase the likelyhood of a race condition.
+                int res = sharedA1.x & sharedB1.x & sharedC1_1.x & sharedC1_1.y & sharedC1_2.x & ((C1)sharedC1_2).y &
+                          sharedA2.x & sharedB2.x & sharedC2_1.x & sharedC2_1.y & sharedC2_2.x & ((C2)sharedC2_2).y &
+                          sharedD1_1.x & ((D1)sharedD1_2).x & sharedD2_1.x & ((D2)sharedD2_2).x;
+                if (res != 1) {
+                    System.err.println("Incorrect field value observed!");
+                    System.exit(1);
```

Comment:
Is there any particular reason why you don't throw a `RuntimeException` here instead?

Reply:
Yes, it's the best way (I think) to make sure all threads immediately exit if one thread observes an error. Throwing an exception will not stop the other threads.

Reply:
Got it, thanks for the explanation!